### PR TITLE
Signal generator now supports more signal types

### PR DESCRIPTION
**Pull request by bminer**
New signal types include square and triangle waves along with random walk

Options for signal generator have changed:
- added signal type option (sine, square, triangle, or random walk)
- amplitude was replaced by min/max value
- offset has been replaced by initial value (for sine, square, and triangle, this is measured in radians)
- added "round to" option. Can round to nearest tenth by setting to 0.1, for example.
- add min/max increment value for "random walk". If you specify -0.1 and 0.2 respectively, the value will tend to increase more than decrease, for example.

Also improved how batch period timer works.

### CODE REVIEW

**Comment by bminer** [comment]
I considered making min/max increment values a percentage rather than a raw value. I also considered just having a single variable called IncrementValue rather than min/max. But then, I discovered an interesting property of using min/max. Imagine a signal generator with these settings:

- Max. value = 1
- Min. value = 0
- Initial value = 0
- Min. increment = -7
- Max. increment = 3
- Round to = 1
- Sample Rate = 20mins

Due to min/max/round to options, this is a binary value, either 0 or 1, biased toward 0. This could be useful for simulating binary switches or something like it. Effectively, this will hold the value for at least 20m and picks a random number between -7 and 3. Due to rounding, if value is currently 0, there's a 25% chance it becomes 1. If 1, there's a 65% chance it becomes 0. This means that the value will be 0 roughly 91.25% (= 75% + (1 - 75%) * 65%) of the time.

So, for this reason, I kept the min/max increment concept.

File: client/signal-generator.go [hunk 1]
```diff
@@ -12,129 +14,283 @@ import (
...
 	stop          chan struct{}
 	newPoints     chan NewPoints
 	newEdgePoints chan NewPoints
-	natsSubject   string
 }
 
 // NewSignalGeneratorClient ...
 func NewSignalGeneratorClient(nc *nats.Conn, config SignalGenerator) Client {
 	return &SignalGeneratorClient{
+		log:           log.New(os.Stderr, "signalGenerator: ", log.LstdFlags|log.Lmsgprefix),
```
Comment (cbrake) [inline client/signal-generator.go:79]: I like the custom logger -- we should start doing this in all the clients/subsystems.

File: client/signal-generator.go [hunk 2]
```diff
@@ -12,129 +14,283 @@ import (
...
 
+/* TODO: Optimization
+
+Note that future designs may keep track of all running SignalGeneratorClients
+and manage only a single batch period timer that runs at a scheduled time.
+One way to do this is to keep track of each clients' next scheduled run time in
+a sorted list. Then, we schedule the timer to run at the soonest scheduled time.
+At that time, we process all clients' batches of points and reschedule another
+timer.
+
```
Comment (cbrake) [inline client/signal-generator.go:57]: Timers are a very useful concept in rule logic as well, so perhaps a common timer node type could be leveraged for this.

**Comment by cbrake** [comment]
Looks great -- nice work!

can you get `siot_test` to pass yet? I'm seeing an elm-review issue:

```
-- ELM-REVIEW ERROR ---------------- src/Components/NodeSignalGenerator.elm:21:9

(fix) NoPrematureLetComputation: Let value was declared prematurely

21|         signalType =
            ^^^^^^^^^^
22|             Point.getText o.node.points Point.typeSignalType ""

This value is only used in some code paths, and it can therefore be computed
unnecessarily.

Try moving it closer to where it is needed, I recommend to move it to line 57.

Errors marked with (fix) can be fixed automatically using `elm-review --fix`.
```

I think you can run `siot_frontend_fix` to fix these issues.

**Review by cbrake — APPROVED** [review]

**Comment by bminer** [comment]
Closing this for now. Will re-open new PR since this one was mostly merged into master.

**Pull request closed without merging**